core/src/main/scala/org/http4s/internal/CharPredicate.scala

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright 2009-2019 Mathias Doenitz
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.internal
+
+import scala.annotation.tailrec
+import scala.collection.immutable.NumericRange
+
+sealed abstract class CharPredicate extends (Char => Boolean) {
```

Comment:
We have decoders based on `CharPredicate`, and they're going to be faster than going char-by-char with cats.parse when we're decoding raw strings.  We could probably whittle this down, but the fastest way to get started was to just take it from parboiled2.